src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -1657,6 +1658,9 @@ void MarkField (FieldDefinition field, in DependencyInfo reason)
 				}
 			}
 
+			if (reason.Kind is DependencyKind.FieldAccess && field.IsStatic && Annotations.TryGetLinkerAttribute (field.DeclaringType, out RequiresUnreferencedCodeAttribute requiresUnreferencedCode))
+				ReportRequiresUnreferencedCode (field.GetDisplayName (), requiresUnreferencedCode, methodCaller);
```

Comment:
This needs to happen before the call to `CheckProcessed` since we want to report the warning on every "call site", not just on the first one.

I played with this some more: https://github.com/tlakollo/linker/pull/1 - please take a look - fixed this issue and added more tests.

Note that I think this will need to warn on almost all Kind - reflection access, DAM access, ...